### DISCUSSION TITLE
Want to extend ModbusTcpSlave class

### DISCUSSION THREAD

**Discussion by Hazelton42** [digitalpetri/modbus#62] [General]
I was looking to extend the ModbusTcpSlave class with a custom bind handler.   I ended up having to pretty much duplicate the whole thing instead of making a subclass because there are a few methods that are private instead of protected.  Could you either make the ModbusTcpSlaveHandler inner class protected or make the onChannelRead, onChannelInactive and onExceptionCaught methods protected instead of private?

**Comment by kevinherron** [comment]
Sure, but can you give me a little more detail on what you were trying to accomplish and why? I'm close to releasing a 2.0 version of this library and want to understand if it's a use case I missed.

**Reply by Hazelton42** [reply]
I work for a company named NetFoundry and we have an open source secure network overlay called "OpenZiti".  Basically it adds secure networking and dark services.   I wanted to replace the `bind` method with a custom one that binds to a ziti service.  This is what I was trying to limit class to be but the `ModbusTcpSlaveHandler` inner class is private:
```
  public ModbusZitiSlave(ModbusTcpSlaveConfig config) {
    super(config);
    this.config = config;
  }
  
  public CompletableFuture<ModbusZitiSlave> bindZiti(ZitiContext zitiContext, String zitiService) {
    CompletableFuture<ModbusZitiSlave> bindFuture = new CompletableFuture<>();
    ChannelInitializer<ZitiChannel> initializer = new ChannelInitializer<>() {
      protected void initChannel(ZitiChannel channel) {
        channelCounter.inc();
        logger.info("channel initialized: {}", channel);
        channel.pipeline().addLast(new LoggingHandler(LogLevel.TRACE));
        channel.pipeline().addLast(new ModbusTcpCodec(new ModbusResponseEncoder(), new ModbusRequestDecoder()));
        channel.pipeline().addLast(new ModbusZitiSlave.ModbusTcpSlaveHandler(ModbusZitiSlave.this));
        channel.closeFuture().addListener(future -> ModbusZitiSlave.this.channelCounter.dec());
      }
    };
    ServerBootstrap bootstrap = new ServerBootstrap();
    this.config.getBootstrapConsumer().accept(bootstrap);
    bootstrap.group(this.config.getEventLoop())
        .channelFactory(new ZitiServerChannelFactory(zitiContext))
        .handler(new LoggingHandler(LogLevel.DEBUG))
        .childHandler(initializer)
        .option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
        .bind(new ZitiAddress.Bind(zitiService)).addListener(future -> {
          if (future.isSuccess() && future instanceof ChannelFuture) {
            Channel channel = ((ChannelFuture) future).channel();
            this.serverChannels.put(channel.localAddress(), channel);
            bindFuture.complete(this);
          } else {
            logger.error("Failed bind {}", future.cause().getMessage());
            bindFuture.completeExceptionally(future.cause());
          }
        });
    return bindFuture;
  }
```

**Reply by kevinherron** [reply]
Ok thanks. I don't mind making the visibility changes. In 2.0 I've pulled out the transport into its own abstraction, so I think you'd end up just creating your own `ZitiServerTransport` class.

```java
public interface ModbusServerTransport<T> {

  /**
   * Bind the transport to its configured local address.
   *
   * @return a {@link CompletionStage} that completes when the transport is bound.
   */
  CompletionStage<Void> bind();

  /**
   * Unbind the transport from its configured local address.
   *
   * @return a {@link CompletionStage} that completes when the transport is unbound.
   */
  CompletionStage<Void> unbind();

  /**
   * Configure a callback to receive request frames from the transport.
   *
   * @param frameReceiver the callback to receive request frames.
   */
  void receive(FrameReceiver<T> frameReceiver);

  interface FrameReceiver<T> {

    /**
     * Receive a request frame from the transport and respond to it.
     *
     * @param frame the request frame.
     * @return a corresponding response frame.
     * @throws Exception if there is an unrecoverable error and the channel should be closed.
     */
    T receive(T frame) throws Exception;

  }

}
```

```java
/**
 * Modbus/TCP server transport; a {@link ModbusServerTransport} that sends and receives
 * {@link ModbusTcpFrame}s.
 */
public interface ModbusTcpServerTransport extends ModbusServerTransport<ModbusTcpFrame> {}
```

**Reply by Hazelton42** [reply]
Great!  Thanks for the update on 2.0 changes.

**Reply by kevinherron** [reply]
@Hazelton42 are these visibility changes enough for now? https://github.com/digitalpetri/modbus/pull/63

**Reply by Hazelton42** [reply]
Looks good.  I was playing around this morning and the only other modification I played around with was adding this:
```
  protected void putServerChannel(SocketAddress socketAddress, Channel channel) {
    this.serverChannels.put(socketAddress, channel);
  }
```
That way I don't have to manage the channels myself and override the shutdown to clean them up.  I can just call `putServerChannel(channel.localAddress(), channel)` in my bind method

**Reply by Hazelton42** [reply]
I pulled that branch and it is working as I wanted

**Reply by ahazeltonNF** [reply]
When can we get this merged and released?

**Reply by kevinherron** [reply]
Ahh, crap, I should have done it last week. 

I'm on a work trip this week and won't have access to the signing keys needed to publish a release until I get back this weekend.

I think this project pre-dates GH actions... so no CI set up to build/publish. Fixed that already for 2.0.

**Reply by kevinherron** [reply]
@Hazelton42 this should be available in Maven Central whenever Sonatype OSS syncs next.

**Reply by ahazeltonNF** [reply]
Working great!  Thanks for the merge and release!
